### DISCUSSION TITLE
Monitor in and out CC Messages

### DISCUSSION THREAD

**Discussion by schoko11** [djipco/webmidi#373] [Q&A]
Hi,

thanks for this great library. 

What i don't understand is why i can't monitor outgoing (and maybe also incoming) CC''s. It is a typescript node app bundled for the browser.

```
import {WebMidi} from "webmidi";
....
midiOutput = WebMidi.getOutputByName(midiOutName!);
midiOutput.channels[1].sendControlChange(CCNumber,1,[1]) ;
```

As hardware i have a Kemper Amplifier, in the Manual is stated that a icon would be shown if messages are received.
When sending this the so called rig, changes as expected, but no icon is shown, and i also can't see any CC Messages on the Midi Connection... (using midisnoop on linux)?

Why is this important ? 
Because i am doing just a timeout after sending controlchange to know when the device did switch to a new (so called) "rig". Now i just don't know when this message was sent. (and another thing i have to ask in the forum of the device: normally the device can reflect the actions as midi, so i can wait for a specific message -> in this example the sent Control Change).
Afterwards I request various other parameters via SYSEX (which i "await" by timestamps ). 

Thanks!

**Comment by djipco** [comment]
I'm not exactly sure what you are asking. You say "_What i don't understand is why i can't monitor outgoing (and maybe also incoming) CC''s._" 

To that, I respond that you can indeed monitor incoming CC messages. This is done with the `addListener()` method of the `InputChannel` object:

```js
const channel = WebMidi.getInputByName("some_input").channels[1];
channel.addListener("controlchange", e => console.log(e));
```
You cannot monitor outgoing CC messages per se but, since you are the one sending them via your code, you already know when they are going out and are free to do whatever else pleases you.

The `OutputChannel` currently does not dispatch events when you call one of its `send*` methods. This is something that could be added but, as far as I know, this is not something people have been needing.

I hope this helps.

**Reply by schoko11** [reply]
Thanks for your reply!

By monitoring outgoing i ment that i can't see outgoing cc's sent with webmidi in e.g. midimon/midisnoop.

Having an event here would maybe help(but if no one else would need that, it would be not worth the effort) because you could somewhat await that the message was sent, and afterwards request via another midi message (in timeout cycle) if the name (of the so called "rig") has changed. Now i have to do a 
`await timeout..` 
of about 200ms to be sure that the sequence does not get messed up...


Thanks for the snipped for incomming cc's, but i now found out that this is not always emitted by the device itself...

Thanks!

**Comment by djipco** [comment]
> By monitoring outgoing i ment that i can't see outgoing cc's sent with webmidi in e.g. midimon/midisnoop.

I use **Midi Monitor.app** on my Mac and can see outgoing CC messages just fine. There's probably something else at play here.

> Having an event here would maybe help(but if no one else would need that, it would be not worth the effort) because you could somewhat await that the message was sent, and afterwards request via another midi message (in timeout cycle) if the name (of the so called "rig") has changed.

Having `OutputChannel` trigger events when sending outbound messages would not be terribly hard. I'm just curious to know if others would find that useful. Anybody?
